### DISCUSSION TITLE
extrudeRotate behavior is unintuitive

### DISCUSSION THREAD

**Discussion by achirita** [jscad/OpenJSCAD.org#1229] [Ideas]
I recently had to use the extrudeRotate function and noticed that it's behavior is not that intuitive.
I thought I would be able to specify the axis around which the 2D geometry is being rotated in order to obtain the 3D geometry, but noticed that's not the case.
What's even stranger is that if I have a 2D geometry in the XY plane and I use extrudeRotate on it, it gets rotated around the Z axis, which make little sense to me.
It's like the geometry is first rotated around the X axis 90 degrees and then the extrudeRotate operation is performed.
Any thought on this one?

**Comment by hrgdavor** [comment]
Those are valid points, but it is difficult to approve changes that would break, already running scripts based on what is more intuitive. Person who first implemented this had some reason to do it this way.

To fit coding in jscad to my personal preference I create utility functions for my scripts.

If @z3dev can also see value in adding the parameter that changes the extrusion axis , then a PR would be a good next step fwd.

**Reply by achirita** [reply]
Backwards compatibility is a good point. The new parameter could be optional and default to the current behavior, or this could be an improvement idea for the next major release (I think that would be version 3).

**Comment by z3dev** [comment]
@achirita some nice ideas here. thanks.

the extradeRotate functionality was probably created to duplicate OpenSCAD, long long ago.

**Reply by z3dev** [reply]
https://en.m.wikibooks.org/wiki/OpenSCAD_User_Manual/2D_to_3D_Extrusion

And the explanation is long and complicated as well.

**Reply by achirita** [reply]
Thanks for the info, it makes sense. 
Does JSCAD still aim to be compatible with OpenSCAD in terms of API or will it evolve in a different way?
I'm talking about already established CAD functions like fillets, chamfers, shells, etc. I guess a good example would be the operations you can perform in something like Fusion 360, Onshape, or maybe even better opencascade which is another opensource 3D CAD library.

**Reply by z3dev** [reply]
> Does JSCAD still aim to be compatible with OpenSCAD in terms of API or will it evolve in a different way?

JSCAD has already moved beyond OpenSCAD, and compatibility is not considered anymore. The focus is on usability via clean API and JavaScript language features, and performance tuned core functions.

We hope that others will make 'compatibility' layers.

**Reply by z3dev** [reply]
@achirita any chance that some API improvements could be added to V3?
